packages/runtime/test-runtime-utils/src/test/insecureUrlResolverTest.spec.ts

```diff
@@ -32,22 +33,119 @@ describe("Insecure Url Resolver Test", () => {
+        assert.strictEqual(resolvedUrl.url, expectedResolvedUrl, "resolved url is wrong");
+
+        const dataStoreId = "dataStore";
+        const absoluteUrl = await resolver.getAbsoluteUrl(resolvedUrl, dataStoreId);
+
+        const expectedResponseUrl = `${hostUrl}/${tenantId}/${fileName}/${dataStoreId}`;
+        assert.strictEqual(absoluteUrl, expectedResponseUrl, "response url is wrong");
+    });
+
+    it("Test RequestUrl for url with a slash at the end", async () => {
+        const testRequest: IRequest = {
+            url: `https://localhost/${fileName}/`,
+            headers: {},
+        };
+        // Mocking window since the resolver depends on window.location.host
+        if (typeof window === "undefined" && typeof global === "object") {
+            // eslint-disable-next-line dot-notation
+            global["window"] = { location: { host:"localhost" } };
+        }
+        const resolvedUrl = await resolver.resolve(testRequest);
+        ensureFluidResolvedUrl(resolvedUrl);
+
+        const expectedResolvedUrl = `fluid://${new URL(ordererUrl).host}/${tenantId}/${fileName}/`;
+        assert.strictEqual(resolvedUrl.url, expectedResolvedUrl, "resolved url is wrong");
+    });
+
+    it("Test RequestUrl for url with 2 slashes at the end", async () => {
+        const testRequest: IRequest = {
+            url: `https://localhost/${fileName}//`,
+            headers: {},
+        };
+        // Mocking window since the resolver depends on window.location.host
+        if (typeof window === "undefined" && typeof global === "object") {
+            // eslint-disable-next-line dot-notation
+            global["window"] = { location: { host:"localhost" } };
+        }
+        const resolvedUrl = await resolver.resolve(testRequest);
+        ensureFluidResolvedUrl(resolvedUrl);
+
+        const expectedResolvedUrl = `fluid://${new URL(ordererUrl).host}/${tenantId}/${fileName}//`;
+        assert.strictEqual(resolvedUrl.url, expectedResolvedUrl, "resolved url is wrong");
+    });
+
+    it("Test RequestUrl for url with special characters", async () => {
+        const testRequest: IRequest = {
+            url: `https://localhost/${fileName}/!@$%^&*()123/dataStore!@$%^`,
+            headers: {},
+        };
+        // Mocking window since the resolver depends on window.location.host
+        if (typeof window === "undefined" && typeof global === "object") {
+            // eslint-disable-next-line dot-notation
+            global["window"] = { location: { host:"localhost" } };
+        }
+        const resolvedUrl = await resolver.resolve(testRequest);
+        ensureFluidResolvedUrl(resolvedUrl);
 
-        const compUrl = `${hostUrl}/${tenantId}/${fileName}/${dataStoreId}`;
-        assert.equal(response, compUrl, "Url should match");
+        const expectedResolvedUrl = `fluid://${new URL(ordererUrl).host}/${tenantId}/${fileName}/!@$123/dataStore!@$`;
```

Comment:
>`fluid://${new URL(ordererUrl).host}/${tenantId}/${fileName}/!@$123/dataStore!@$`; [](start = 36, length = 82)

This expected URL does not seem right. Why would some  of the special characters be removed? I am guessing its dependent on what this parsing does:
`const parsedUrl = new URL(request.url);`
However, I am not sure what the expected behavior should be. We might have to look into it.

It might also be interesting to see what happens if there are special characters in the "documentId" since it is encoded using "encodeURIComponent".

Reply:
Oh, this got changed while I was reviewing it :-).

---
In reply to: [496196065](https://github.com/microsoft/FluidFramework/pull/3577#discussion_r496196065) [](ancestors = 496196065)